File: frontend/src/api/archive.ts
/**
 * 案卷管理相关 API
 */
import { request, type PaginatedResponse } from './index'

export interface CaseFile {
  id: number
  caseNo: string
  caseName: string
  title: string
  caseType: string
  sourceDepartment: string
  incidentTime?: Date | string
  personName?: string
  personInfo?: {
    gender?: string
    nationality?: string
    birthPlace?: string
    enlistmentTime?: string
    position?: string
    personCategory?: string
  }
  charge?: string
  suicideMethod?: string
  incidentProcess?: string
  investigationProcessAndConclusion?: string
  causeAndLesson?: string
  caseFiling?: string
  judgment?: string
  filePath: string
  fileSize: number
  fileType: string
  ocrText?: string
  metadata?: Record<string, any>
  tags?: string[]
  classificationLevel1?: string
  classificationLevel2?: string
  classificationLevel3?: string
  timeline?: Array<{
    time: string
    event: string
    type: 'incident' | 'investigation' | 'filing' | 'judgment' | 'system'
    typeLabel: string
    description?: string
    timestamp: number
  }>
  status: string
  createdBy?: number
  createdAt: Date | string
  updatedAt: Date | string
}

export interface CaseFileListParams {
  keyword?: string
  caseType?: string
  status?: string
  dateRange?: [Date, Date]
  page?: number
  pageSize?: number
}

/** 案卷导入 SSE 单条事件（四阶段：upload / parse / analyze / complete） */
export interface ImportStreamEvent {
  stage?: 'upload' | 'parse' | 'analyze' | 'complete'
  fileIndex?: number
  fileName?: string
  total?: number
  progress?: number
  success?: boolean
  reason?: string
}

/** 案卷导入 SSE 任务结束事件 */
export interface ImportStreamTaskDone {
  event: 'task_done'
  task_id: number
  total_files: number
  success_files: number
  failed_files: number
  status: string
}

export const caseFileApi = {
  // 获取案卷列表
  getList(params: CaseFileListParams): Promise<PaginatedResponse<CaseFile>> {
    // 将驼峰命名转换为下划线命名，以匹配后端 API
    const queryParams: Record<string, any> = {
      keyword: params.keyword,
      case_type: params.caseType,
      status: params.status,
      page: params.page || 1,
      page_size: params.pageSize || 20
    }
    // 处理日期范围
    if (params.dateRange && params.dateRange.length === 2) {
      queryParams.start_date = params.dateRange[0]
      queryParams.end_date = params.dateRange[1]
    }
    return request.get('/case-file/list', { params: queryParams })
  },
  
  // 获取案卷详情
  getDetail(id: number): Promise<CaseFile> {
    return request.get(`/case-file/detail/${id}`)
  },
  
  // 搜索案卷（全文检索）
  search(params: {
    keyword: string
    searchMode?: 'fuzzy' | 'exact'
    searchScope?: string
    caseType?: string
    department?: string
    sortBy?: 'relevance' | 'time' | 'title'
    page?: number
    pageSize?: number
  }): Promise<PaginatedResponse<{
    id: number
    caseNo: string
    caseName: string
    title: string
    caseType: string
    sourceDepartment: string
    date: string
    relevance: number
    relevanceScore: string
    fragments: string[]
    tags: string[]
  }> & { meta: { took: number; keyword: string } }> {
    // 将参数转换为查询字符串
    const queryParams: Record<string, any> = {
      keyword: params.keyword,
      search_mode: params.searchMode || 'fuzzy',
      search_scope: params.searchScope || 'title,content,metadata,tags',
      sort_by: params.sortBy || 'relevance',
      page: params.page || 1,
      page_size: params.pageSize || 20
    }
    if (params.caseType) queryParams.case_type = params.caseType
    if (params.department) queryParams.department = params.department
    
    return request.post('/case-file/search', null, { params: queryParams })
  },
  
  // 导入案卷（文档上传 → 解析文字 → AI 提取核心内容 → 写入待审核）
  import(formData: FormData, options?: { taskName?: string; sourceDepartment?: string }): Promise<{
    task_id: number
    total_files: number
    success_files: number
    failed_files: number
    status: string
  }> {
    const params: Record<string, string> = {}
    if (options?.taskName) params.task_name = options.taskName
    if (options?.sourceDepartment) params.source_department = options.sourceDepartment
    return request.upload('/case-file/import', formData, { params: Object.keys(params).length ? params : undefined })
  },

  /**
   * 案卷导入（SSE 流式进度）
   * 四个阶段：上传 → 解析 → 智能分析 → 完成
   * @param formData 文件表单
   * @param options 批次名称、来源部门
   * @param onEvent 收到每条 SSE 事件时回调（stage: upload | parse | analyze | complete；event: task_done | error）
   * @returns 流结束后 resolve 最终结果，出错时 reject
   */
  importStream(
    formData: FormData,
    options: { taskName?: string; sourceDepartment?: string },
    onEvent: (data: ImportStreamEvent) => void
  ): Promise<ImportStreamTaskDone> {
    const base = import.meta.env.VITE_API_BASE_URL || '/api/v1'
    const baseUrl = base.startsWith('http') ? base : `${window.location.origin}${base.startsWith('/') ? '' : '/'}${base}`
    const url = new URL('case-file/import/stream', baseUrl.endsWith('/') ? baseUrl : baseUrl + '/')
    if (options.taskName) url.searchParams.set('task_name', options.taskName)
    if (options.sourceDepartment) url.searchParams.set('source_department', options.sourceDepartment)
    const token = localStorage.getItem('token') || sessionStorage.getItem('token')
    return fetch(url.toString(), {
      method: 'POST',
      headers: {
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      body: formData,
    }).then(async (response) => {
      if (!response.ok) {
        const err = await response.json().catch(() => ({}))
        throw new Error((err as any).detail || response.statusText)
      }
      const reader = response.body?.getReader()
      const decoder = new TextDecoder()
      if (!reader) throw new Error('无法读取响应流')
      let buffer = ''
      let taskDone: ImportStreamTaskDone | null = null
      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''
        for (const line of lines) {
          if (!line.startsWith('data: ')) continue
          try {
            const data = JSON.parse(line.slice(6)) as ImportStreamEvent | ImportStreamTaskDone
            if ((data as any).event === 'task_done') {
              taskDone = data as ImportStreamTaskDone
            } else if ((data as any).event === 'error') {
              throw new Error((data as any).message || '导入失败')
            } else {
              onEvent(data as ImportStreamEvent)
            }
          } catch (e) {
            if (e instanceof Error && e.message !== '导入失败') {
              // JSON 解析等错误忽略单条，继续读流
              continue
            }
            throw e
          }
        }
      }
      if (!taskDone) throw new Error('未收到任务完成事件')
      return taskDone
    })
  },
  
  // 获取导入任务列表
  getImportTasks(): Promise<any[]> {
    return request.get('/case-file/import-tasks')
  },
  
  // 删除案卷
  delete(id: number): Promise<void> {
    return request.delete(`/case-file/${id}`)
  },

  /** 根据文档重新生成案卷字段（AI 重新提取），用于卷宗审核页「AI 重新生成」 */
  reExtract(caseFileId: number): Promise<{ extractedData: Record<string, any> }> {
    return request.post(`/case-file/re-extract/${caseFileId}`).then((res: any) => res.data ?? res)
  }
}

File: backend/app/models/archive.py
"""
案卷模型
"""
from sqlalchemy import Column, BigInteger, String, Text, JSON, DateTime, ForeignKey, func, Index
from sqlalchemy.orm import relationship
from app.core.database import Base


class CaseFile(Base):
    """案卷表模型"""
    __tablename__ = "case_files"
    
    id = Column(BigInteger, primary_key=True, autoincrement=True, comment="案卷ID")
    case_no = Column(String(50), unique=True, index=True, comment="案卷编号")
    
    # 核心字段：卷宗名（时间+事发单位-人员类别+姓名+涉案罪名）
    case_name = Column(String(500), index=True, comment="卷宗名: 时间+事发单位-人员类别+姓名+涉案罪名(自杀方式、线索)")
    
    # 核心字段：内容相关
    incident_time = Column(DateTime, index=True, comment="发生时间")
    person_name = Column(String(100), index=True, comment="姓名")
    person_info = Column(JSON, comment="人员基本情况: 性别、民族、出生地、入伍时间、部职别、人员类别")
    charge = Column(String(200), comment="涉案罪名")
    suicide_method = Column(Text, comment="自杀方式或线索（若涉及）")
    incident_process = Column(Text, comment="事发经过")
    investigation_process_and_conclusion = Column(Text, comment="侦查调查过程及结论")
    cause_and_lesson = Column(Text, comment="原因教训")
    case_filing = Column(Text, comment="立案情况")
    judgment = Column(Text, comment="判决情况")
    
    # 原有字段
    title = Column(String(500), comment="标题")
    case_type = Column(String(50), index=True, comment="案卷类型")
    source_department = Column(String(100), comment="来源部门")
    file_path = Column(String(500), comment="文件路径")
    file_size = Column(BigInteger, comment="文件大小(字节)")
    file_type = Column(String(50), comment="文件类型")
    ocr_text = Column(Text, comment="OCR识别文本")
    meta_data = Column(JSON, comment="元数据")  # 使用 meta_data 避免与 SQLAlchemy 的 metadata 保留字段冲突
    tags = Column(JSON, comment="标签")
    classification_level1 = Column(String(50), comment="一级分类")
    classification_level2 = Column(String(50), comment="二级分类")
    classification_level3 = Column(String(50), comment="三级分类")
    timeline = Column(JSON, comment="案件时间线: [{time, event, type, typeLabel, description, timestamp}]")
    status = Column(String(20), default="pending", index=True, comment="状态: pending/processing/completed/failed")
    created_by = Column(BigInteger, ForeignKey("users.id", ondelete="SET NULL"), comment="创建人ID")
    created_at = Column(DateTime, server_default=func.now(), index=True, comment="创建时间")
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now(), comment="更新时间")
    
    # 关系
    creator = relationship("User", back_populates="case_files", foreign_keys=[created_by])
    
    # 全文索引（MySQL 需要单独创建，这里只是标记）
    __table_args__ = (
        Index("idx_fulltext", "case_name", "title", "ocr_text", mysql_prefix="FULLTEXT"),
    )
